### PR TITLE
Fix/ao with zoom

### LINKED ISSUE #303: ambient occlusion changes with zoom
## Description

When zooming in or out, agents get dimmer or brighter and it is not clear why.
We suspect the screen space AO sampling.

## Expected Behavior

Basically want things to stay the same brightness against zooming.

If AO is the problem, consider scaling the spacing of AO samples with depth value.

### PR DESCRIPTION
**Pull request by toloudis**
Problem
=======
Resolves #303 
Ambient Occlusion had poor behavior when zooming in close.  
This may address #306 too.

Because AO is a screen space algorithm here, we are not going to get a perfect result in all cases.   The things that affect the result are generally the range of very small geometries vs very large ones, and overall scene size (size of distance values).  A scene with a very large bounding box and lots of very small geometries can look different than a scene with a small bounding box and just a few objects filling it.

Solution
========
I chose to replace the AO algorithm completely.  The new one uses fewer samples, has more customizable parameters, and has a smoother result.  The quality is good enough to use with one single render pass as opposed to two, which we were using before.

The AO result is not identical to the previous.  Some care should be taken to verify.  I tried to use settings and variables to scale with scene size and distance such that it is reasonably close to our previous production trajectories.

## Type of change

* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)

Screenshots (optional):
-----------------------
The screen shots were not taken at identical viewer sizes so in some cases the number of pixels may also affect the differences.

Before:
Simularium 🔊 2023-05-25 15-09-32
After
Simularium viewer 🔊 2023-05-25 15-09-56

Before:
Simularium 🔊 2023-05-25 15-06-12
After:
Simularium viewer 🔊 2023-05-25 15-08-24

Before:
Simularium 🔊 2023-05-25 15-05-29
After:
Simularium viewer 🔊 2023-05-25 15-08-55

### CODE REVIEW

File: examples/Viewer.tsx [hunk 1]
```diff
@@ -127,7 +127,7 @@ const initialState = {
     currentFrame: 0,
     currentTime: 0,
     height: 700,
-    width: 800,
+    width: 1200,
```
Comment (toloudis) [inline examples/Viewer.tsx:130]: widening the example viewer allows me to use the ctrl-alt-1 advanced settings without hiding so much of the view.

File: examples/Viewer.tsx [hunk 2]
```diff
@@ -547,33 +547,26 @@ class Viewer extends React.Component<{}, ViewerState> {
...
-                    <option value="ATPsynthase_10.h5">ATP 10</option>
+                    <option value="nanoparticle_wrapping.simularium">Nanoparticle wrapping</option>
+                    <option value="smoldyn_min1.simularium">Smoldyn min1</option>
+                    <option value="smoldyn_spine.simularium">Smoldyn spine</option>
+                    <option value="medyan_Chandrasekaran_2019_UNI_alphaA_0.1_MA_0.675.simularium">medyan 625</option>
+                    <option value="medyan_Chandrasekaran_2019_UNI_alphaA_0.1_MA_0.0225.simularium">medyan 0225</option>
+                    <option value="springsalad_condensate_formation_Below_Ksp.simularium">springsalad below ksp</option>
+                    <option value="springsalad_condensate_formation_At_Ksp.simularium">springsalad at ksp</option>
+                    <option value="springsalad_condensate_formation_Above_Ksp.simularium">springsalad above ksp</option>
+                    <option value="blood-plasma-1.0.simularium">blood plasma</option>
```
Comment (toloudis) [inline examples/Viewer.tsx:569]: vastly better test data

File: src/simularium/RemoteSimulator.ts [hunk 3]
```diff
@@ -156,17 +156,17 @@ export class RemoteSimulator implements ISimulator {
...
         this.webSocketClient.addJsonMessageHandler(
             NetMessageEnum.ID_UPDATE_RATE_PARAM,
-            this.updateRateParam
+            (_msg) => this.updateRateParam()
         );
 
         this.webSocketClient.addJsonMessageHandler(
             NetMessageEnum.ID_MODEL_DEFINITION,
-            this.onModelDefinitionArrive
+            (_msg) => this.onModelDefinitionArrive()
```
Comment (toloudis) [inline src/simularium/RemoteSimulator.ts:169]: changes in this file are a bonus bug fix - the heartbeat ping was raising exceptions in react and breaking the viewer, not sure when bug was introduced.

File: src/visGeometry/rendering/SimulariumRenderer.ts [hunk 4]
```diff
@@ -194,59 +181,26 @@ class SimulariumRenderer {
...
+            max: 0.1,
+        });
+
         const depth = gui.addFolder({ title: "DepthCueing", expanded: false });
-        depth.addInput(settings, "atomBeginDistance", { min: 0.0, max: 300.0 });
+        depth.addInput(settings, "atomBeginDistance", { min: 0.0, max: 2.0 });
         depth.addInput(settings, "chainBeginDistance", {
             min: 0.0,
-            max: 300.0,
+            max: 2.0,
```
Comment (toloudis) [inline src/visGeometry/rendering/SimulariumRenderer.ts:203]: these distances are now percentages of the bounding box size

File: src/visGeometry/rendering/CompositePass.ts [hunk 5]
```diff
@@ -158,57 +156,72 @@ class CompositePass {
...
-                    // near is atomColor, far is color.xyz
-                    // linear RGB interp? not HCL?
-                    color.xyz = mix(color.xyz, atomColor, t);
-                    //color.xyz = atomColor;
-                    //color.xyz = vec3(0.0, 1.0, 0.0);
-                }
-            
-                gl_FragColor = vec4( occ1 * occ2 * color.xyz, 1.0);
+                // color = max(color, vec3(0.0,0.0,0.0));
+                // color = min(color, vec3(1.0,1.0,1.0));
```
Comment (frasercl) [inline src/visGeometry/rendering/CompositePass.ts:205]: these lines are probably not worth keeping around as comments

**Review by frasercl — APPROVED** [review]
LGTM - couple suggestions but no problems

File: src/visGeometry/rendering/SimulariumRenderer.ts [hunk 6]
```diff
@@ -382,19 +333,41 @@ class SimulariumRenderer {
...
+        this.ssao1Pass.pass.material.uniforms.minResolution.value =
+            this.parameters.ao1.minResolution;
+        this.ssao1Pass.pass.material.uniforms.beginFalloffDistance.value =
+            this.parameters.atomBeginDistance * sceneSize +
+            Math.max(this.boundsNear, 0.0);
+        this.ssao1Pass.pass.material.uniforms.endFalloffDistance.value =
+            this.parameters.chainBeginDistance * sceneSize +
+            Math.max(this.boundsNear, 0.0);
+
+        this.ssao1Pass.pass.material.uniforms.cameraFar.value = camera.far;
```
[the remark this answers is not in the document]
Comment (toloudis) [reply src/visGeometry/rendering/SimulariumRenderer.ts:357]: I agree, it's awkward accessing properties 6 levels deep.  Maybe a separate PR to just add that function...

**Review by blairlyons — APPROVED** [review]

**Pull request merged**